analyzers/tests/SonarAnalyzer.UnitTest/TestFramework/SymbolicExecution/SymbolicValueAssertions.cs

```diff
@@ -0,0 +1,164 @@
+                .FailWith("Expected {context:symbolicValue} to have only constraint {0}{reason}, but symbolicValue has {1} constraints.",
+                    _ => constraint,
+                    x => x.OrderBy(x => x.ToString()))
+                .Then
+                .ForCondition(allConstraints => allConstraints[0] == constraint)
+                .FailWith("Expected {context:symbolicValue} to have constraint {0}{reason}, but symbolicValue has {1} constraint.",
+                    _ => constraint,
+                    allConstraints => allConstraints[0]);
+            var matchedConstrainted = Subject.AllConstraints.First(x => x == constraint);
+            return new(this, matchedConstrainted);
+        }
+
+        public AndConstraint<SymbolicValueAssertions> HaveNoConstraints(string because = "", params object[] becauseArgs)
+        {
+            Execute.Assertion
+                .BecauseOf(because, becauseArgs)
+                .ForCondition(Subject == null || Subject.AllConstraints.All(x => x == null))
+                .FailWith("Expected {context:symbolicValue} to have no constraints{reason}, but {0} constraints were found.", Subject?.AllConstraints.OrderBy(x => x.ToString()));
+            return new(this);
+        }
+
+        public AndConstraint<SymbolicValueAssertions> HaveConstraints(params SymbolicConstraint[] constraints)
+            => HaveConstraints(constraints, string.Empty);
+
+        public AndConstraint<SymbolicValueAssertions> HaveConstraints(IEnumerable<SymbolicConstraint> constraints, string because = "", params object[] becauseArgs)
+        {
+            HaveConstraintsCommonAssertions(constraints, because, becauseArgs)
+                .Given(allConstraints => constraints.Except(allConstraints).OrderBy(x => x.ToString()).ToList())
+                .ForCondition(missingConstraints => missingConstraints.Count == 0)
+                .FailWith("Expected {context:symbolicValue} to have constraints {0}{reason}, but constraints {1} are missing. Actual constraints {2}.",
+                    _ => constraints,
+                    x => x,
+                    _ => Subject.AllConstraints.OrderBy(x => x.ToString()));
+            return new(this);
+        }
+
+        public AndConstraint<SymbolicValueAssertions> HaveOnlyConstraints(params SymbolicConstraint[] constraints)
+            => HaveOnlyConstraints(constraints, string.Empty);
+
+        public AndConstraint<SymbolicValueAssertions> HaveOnlyConstraints(IEnumerable<SymbolicConstraint> constraints, string because = "", params object[] becauseArgs)
+        {
+            HaveConstraintsCommonAssertions(constraints, because, becauseArgs)
+                .Given(allConstraints => new
+                {
+                    missingConstraints = constraints.Except(allConstraints).OrderBy(x => x.ToString()).ToList(),
+                    addionalConstraints = allConstraints.Except(constraints).OrderBy(x => x.ToString()).ToList()
+                })
```

Comment:
I struggle to follow this logic. Let's try to rename the stuff to make it easier to understand what is what.

We know that all of them are constraints from their types. I need to better see what they really mean.
```suggestion
        public AndConstraint<SymbolicValueAssertions> HaveOnlyConstraints(IEnumerable<SymbolicConstraint> expected, string because = "", params object[] becauseArgs)
        {
            HaveConstraintsCommonAssertions(constraints, because, becauseArgs)
                .Given(actual => new
                {
                    missing = expected.Except(actual).OrderBy(x => x.ToString()).ToList(),
                    addional = actual.Except(expected).OrderBy(x => x.ToString()).ToList()
                })
```

Reply:
Same applies everywhere else